phase4-entsog-client/src/main/java/com/helger/phase4/entsog/Phase4ENTSOGSender.java

```diff
@@ -0,0 +1,191 @@
+
+import com.helger.phase4.attachment.Phase4OutgoingAttachment;
+import com.helger.phase4.attachment.WSS4JAttachment;
+import com.helger.phase4.client.AS4ClientUserMessage;
+import com.helger.phase4.crypto.ECryptoKeyIdentifierType;
+import com.helger.phase4.sender.AS4BidirectionalClientHelper;
+import com.helger.phase4.sender.AbstractAS4UserMessageBuilder;
+import com.helger.phase4.util.AS4ResourceHelper;
+import com.helger.phase4.util.Phase4Exception;
+
+/**
+ * This class contains all the specifics to send AS4 messages with the ENTSOG
+ * profile. See <code>sendAS4Message</code> as the main method to trigger the
+ * sending, with all potential customization.
+ *
+ * @author Pavel Rotek
+ * @since 0.13.3
+ */
+@Immutable
+public final class Phase4ENTSOGSender {
+	private static final Logger LOGGER = LoggerFactory.getLogger(Phase4ENTSOGSender.class);
+
+	private Phase4ENTSOGSender() {
+	}
+
+	/**
+	 * @return Create a new Builder for AS4 messages if the payload is present.
+	 *         Never <code>null</code>.
+	 */
+	@Nonnull
+	public static ENTSOGUserMessageBuilder builder() {
+		return new ENTSOGUserMessageBuilder();
+	}
+
+	/**
+	 * Abstract ENTSOG UserMessage builder class with sanity methods
+	 *
+	 * @author Philip Helger
+	 * @param <IMPLTYPE> The implementation type
+	 */
+	public abstract static class AbstractENTSOGUserMessageBuilder<IMPLTYPE extends AbstractENTSOGUserMessageBuilder<IMPLTYPE>>
+			extends AbstractAS4UserMessageBuilder<IMPLTYPE> {
+
+		private ECryptoKeyIdentifierType m_signingKeyIdentifierType = ECryptoKeyIdentifierType.ISSUER_SERIAL;
+		private ECryptoKeyIdentifierType m_encryptionKeyIdentifierType = ECryptoKeyIdentifierType.ISSUER_SERIAL;
+		
+		private Phase4OutgoingAttachment m_aPayload;
+		private Charset m_aPayloadCharset;
+		private String m_aPayloadDocumentType;
+		
+		protected AbstractENTSOGUserMessageBuilder() {
+			// Override default values
+			try {
+				httpClientFactory(new Phase4ENTSOGHttpClientSettings());
+			} catch (final Exception ex) {
+				throw new IllegalStateException("Failed to init AS4 Client builder", ex);
+			}
+		}
+
+		/**
+		 * Encryption Key Identifier Type.
+		 *
+		 * @param keyIdentifier ECryptoKeyIdentifierType. Defaults to ISSUER_SERIAL
+		 * @return this for chaining
+		 */
+		public final IMPLTYPE setEncryptionKeyIdentifierType(ECryptoKeyIdentifierType encryptionKeyIdentifierType) {
+			m_encryptionKeyIdentifierType = encryptionKeyIdentifierType;
+			return thisAsT();
+		}
+
+		/**
+		 * Signing Key Identifier Type.
+		 *
+		 * @param keyIdentifier ECryptoKeyIdentifierType. Defaults to ISSUER_SERIAL
+		 * @return this for chaining
+		 */
+		public final IMPLTYPE setSigningKeyIdentifierType(ECryptoKeyIdentifierType signingKeyIdentifierType) {
+			m_signingKeyIdentifierType = signingKeyIdentifierType;
+			return thisAsT();
+		}		
+		
+		/**
+		 * Set the payload to be send out.
+		 *
+		 * @param aBuilder The payload builder to be used.
+		 * @return this for chaining
```

Comment:
The pararameter javadocs for charset and documentType are missing. What exactly is the "document type"?

Reply:
Charset and Document Type are both parameters for PayloadInfo.
 
DocumentType is ENSOG specific parameter that carries semantic EDIGAS type of the document (eg Nomination can have value 01G). This type can be later used for eg. for XSD validation decission.

Charset is another recommended attribute in PayloadInfo and means charset of the transferred payload. It could be potentially set on Phase4OutgoingAttachment, but there is no suitable attribute. I can imagine this param can be used in other profiles too. eg upcomming ENTSOE. I feel the charset attribute should be encapsulated directly in Phase4OutgoingAttachment.

Reply:
Yes, I agree the `Charset` parameter should go into `Phase4OutgoingAttachment` - on it.